docs/developers/reproducible-contract-builds.md

```diff
@@ -0,0 +1,203 @@
+
+:::important
+As of September 2022, the Rust toolchain does not support reproducible builds out-of-the-box, thus we recommend smart contract developers to follow this tutorial in order to achieve deterministic compilation.
+:::
+
+## Smart Contract `codehash`
+
+Before diving into contract build reproducibility, let's grasp the concept of `codehash`.
+
+When a smart contract is deployed, the network stores the bytecode, and also computes its `blake2b` checksum (using a digest length of 256 bits). This is called the `codehash`.
+
+Assume that we are interested into the following contract, deployed on _devnet_: [erd1qqqqqqqqqqqqqpgqahertgz4020wegswus8m7f2ak8a6d0gv396qw3t2zy](https://devnet-explorer.elrond.com/accounts/erd1qqqqqqqqqqqqqpgqahertgz4020wegswus8m7f2ak8a6d0gv396qw3t2zy).
+
+Using Python3 and [`requests`](https://requests.readthedocs.io) (for readability), we can fetch its _codehash_ from the API:
+
+```
+import base64
+
+import requests
+
+contract = "erd1qqqqqqqqqqqqqpgqahertgz4020wegswus8m7f2ak8a6d0gv396qw3t2zy"
+data = requests.get(f"https://devnet-api.elrond.com/accounts/{contract}").json()
+
+# The API returns the base-64 encoded codehash
+codehash_base64 = data["codeHash"]
+codehash = base64.b64decode(codehash_base64)
+codehash_hex = codehash.hex()
+
+print(codehash_hex)
+```
+
+The output is:
+
+```
+8da80dd8bf5b32dfdeaa919901c679d8167162aed7eb7ea41502c7f3d1b27e25
+```
+
+Alternatively, we can locally compute the _codehash_, as well (using the `blake2b` module):
+
+```
+from hashlib import blake2b
+
+import requests
+
+contract = "erd1qqqqqqqqqqqqqpgqahertgz4020wegswus8m7f2ak8a6d0gv396qw3t2zy"
+data = requests.get(f"https://devnet-api.elrond.com/accounts/{contract}").json()
+code_hex = data["code"]
+code = bytes.fromhex(code_hex)
+
+h = blake2b(digest_size=32)
+h.update(code)
+codehash_hex = h.hexdigest()
+
+print(codehash_hex)
+```
+
+If the `WASM` file is directly available, the snippet simplifies to:
+
+```
+from hashlib import blake2b
+
+bytecode_file = open("adder.wasm", "rb")
+code = bytecode_file.read()
+bytecode_file.close()
+
+h = blake2b(digest_size=32)
+h.update(code)
+codehash_hex = h.hexdigest()
+
+print(codehash_hex)
+```
+
+We can also use the utility `b2sum` to locally compute the _codehash_:
+
+```
+b2sum -l 256 adder.wasm
+```
+
+Now, it's easy to verify the bytecode equality of two given builds of a contract using solely the _codehash_ property.
+
+## Supporting reproducible builds
+
+As of September 2022, the recommended approach to support reproducible builds for your smart contract is to perform the **deployment build** (as opposed to regular development builds) in a Docker container based on a [publicly-available, tagged Docker image](https://hub.docker.com/r/elrondnetwork/elrond-sdk-erdpy-rust/tags), that includes tagged, explicit versions of the build tools (_erdpy_, _Rust_, _wasm-opt_ etc.). In addition, the current directory when running the build command should be: `/home/elrond`. 
```

Comment:
The `/home/elrond` current dir is already set in the docker image, right? If so, I would specify it as a less important note that it should not be changed. This phrase could make devs ask themselves: "hm, how can I set the current dir?"

Reply:
Right, fixed.